File: lclang/src/main/java/lclang/statements/BlockStatement.java
package lclang.statements;

import lclang.Caller;
import lclang.LCBaseExecutor;
import lclang.Link;
import lclang.exceptions.LCLangRuntimeException;
import lclang.libs.lang.classes.VoidClass;

public class BlockStatement extends Statement {
    public final Statement[] statements;

    public BlockStatement(Statement[] statements) {
        super(0);
        this.statements = statements;
    }

    @Override
    public Link visit(Caller prevCaller, LCBaseExecutor visitor) throws LCLangRuntimeException {
        for(Statement stmt: statements) {
            Caller caller = stmt.getCaller(prevCaller);
            Link value = stmt.visit(caller, visitor);
            if (value.state!= Link.State.NOTHING)
                return value;
            else value.get.invoke(caller);
        }

        return VoidClass.value;
    }
}

File: lclang/src/main/kotlin/lclang/statements/BlockStatement.kt
package lclang.statements

import lclang.Caller
import lclang.LCBaseExecutor
import lclang.VOID_VALUE
import lclang.Value

class BlockStatement(
    val statements: Array<Statement>,
    line: Int,
    column: Int
): Statement(line, column) {
    override fun visit(prevCaller: Caller, visitor: LCBaseExecutor): Value {
        for(stmt in statements) {
            val caller = stmt.getCaller(prevCaller)
            val value = stmt.visit(caller, visitor)
            if (value.state!= Value.State.NOTHING)
                return value
            else value.get(caller)
        }

        return VOID_VALUE
    }
}